modules/gapi/src/streaming/onevpl/accelerators/utils/shared_lock.cpp

```diff
@@ -0,0 +1,96 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2021 Intel Corporation
+
+#include <thread>
+#include "streaming/onevpl/accelerators/utils/shared_lock.hpp"
+
+
+namespace cv {
+namespace gapi {
+namespace wip {
+namespace onevpl {
+
+SharedLock::SharedLock() {
+    exclusive_lock.store(false);
+    shared_counter.store(0);
+}
+
+size_t SharedLock::shared_lock() {
+    size_t prev = 0;
+    bool in_progress = false;
+    bool pred_excl = exclusive_lock.load();
+    do {
+        if (!pred_excl) {
+            // if no exclusive lock than start shared lock transaction
+            prev = shared_counter.fetch_add(1);
+            in_progress = true; //transaction is in progress
+        } else {
+            if (in_progress) {
+                in_progress = false;
+                shared_counter.fetch_sub(1);
+            }
+            std::this_thread::yield();
+        }
+
+        // test if exclusive lock happened before
+        pred_excl = exclusive_lock.load();
+    } while (pred_excl || !in_progress);
+
+    return prev;
+}
+
+size_t SharedLock::unlock_shared() {
+    return shared_counter.fetch_sub(1);
+}
+
+void SharedLock::lock() {
+    bool in_progress = false;
+    size_t prev_shared = shared_counter.load();
+    do {
+        if (prev_shared == 0) {
+            bool expected = false;
+            while (!exclusive_lock.compare_exchange_strong(expected, true)) {
+                expected = false;
+                std::this_thread::yield();
+            }
+            in_progress = true;
+        } else {
+            if (in_progress) {
+                in_progress = false;
+                exclusive_lock.store(false);
+            }
+            std::this_thread::yield();
+        }
+        prev_shared = shared_counter.load();
+    } while (prev_shared != 0 || !in_progress);
```

Comment:
That's what I meant - in the end, `in_progress == true`, but you must proceed further. The comment is regarding naming. "In progress" itself means you are in the middle of some process and you cannot move further. It doesn't match with the meaning of this variable

If to suggest, this variable means that write access is gained, so it can be named like `gained`, `succeeded`; or, better, meaningful `locked`, because `exclusive_lock` **was** set true, but in the case you've described it should be rolled back. But it is `locked` that time.

Reply:
this variable cannot be named as `gained`, `succeeded` of `locked` because resource is not locked until we finished final `while`. It actually means: `we are in the middle of progress something` and we must not make  some assumption about succeed or not until we check the second condition  `shared_lock ==0`

So `in_progress ` means: I increased exclusive lock counter but i'm not sure that i'm allowed to continue with exclusive access until i check the second condition.
This is sync with using two independent atomic states: and we cannot proceed it both of them wouldn't reach the right state